### PR TITLE
C++: Implement proper coroutine support in IR

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR fixes the IR generation for `co_await` and `co_yield`. We added IR support for these in https://github.com/github/codeql/pull/16187. However, we didn't quite capture the right semantics of the constructs. This PR fixes this by generating the (hopefully) correct semantics by making use of the children of `co_await` and `co_yield` as emitted by our frontend.

This gets rid of all the dead ends in the IR. I haven't quite figured out why we get new `nonUniqueIRVariable` inconsistencies, though (and whether this is a QL or extractor bug).

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll [hunk 1]
```diff
@@ -1301,19 +1299,153 @@ class TranslatedUnaryExpr extends TranslatedSingleInstructionExpr {
...
 
+/**
+ * IR translation of a `co_await` or `co_yield` expression.
+ *
+ * The translation of `x = co_await ...` is essentially:
+ * ```cpp
+ * if !awaiter.await_ready() {
+ *   awaiter.await_suspend()
+ * }
+ * x = awaiter.await_resume();
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll:1317]:
```suggestion
 * if (!awaiter.await_ready()) {
 *   awaiter.await_suspend();
 * }
 * x = awaiter.await_resume();
```

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/InstructionTag.qll [hunk 2]
```diff
@@ -186,6 +187,8 @@ string getInstructionTagId(TInstructionTag tag) {
   or
   tag = BoolConversionCompareTag() and result = "BoolConvComp"
   or
+  tag = ResultCopyTag() and result = "ResultCopy"
+  or
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/InstructionTag.qll:191]: This was a drive-by fix I did while debugging. This tag didn't have a result previously.

**Comment by jketema** [comment]
> This gets rid of all the dead ends in the IR. I haven't quite figured out why we get new `nonUniqueIRVariable` inconsistencies, though (and whether this is a QL or extractor bug).

I think this is related to the issue I reported to our frontend provider (but let me double check that).

**Comment by MathiasVP** [comment]
> I think this is related to the issue I reported to our frontend provider (but let me double check that).

Awesome. That will save me half a morning of debugging then 🎉

**Comment by jketema** [comment]
> > I think this is related to the issue I reported to our frontend provider (but let me double check that).
> 
> Awesome. That will save me half a morning of debugging then 🎉

Yeah, those are exactly the local variables that we filed an issue for with our frontend provider.

If these are not causing any serious problems, I propose with just add a comment to the open issue on this that once we receive a patch we should also fix this (if that's not automatic).

**Comment by jketema** [comment]
One question: how do the translations of the reuse expression look (they are a bit hard to find)? In the intermediate state we had before this PR, the use and reuse we in separate blocks. Is that solved now?

**Comment by jketema** [comment]
> One question: how do the translations of the reuse expression look (they are a bit hard to find)? In the intermediate state we had before this PR, the use and reuse we in separate blocks. Is that solved now?

Ah, they still seem to be in different block, but the order of the blocks seems such that we get something that is initialised.

**Review by jketema — APPROVED** [review]

**Comment by MathiasVP** [comment]
> Ah, they still seem to be in different blocks, but the order of the blocks seems such that we get something that is initialised.

Indeed. It seems like the frontend is doing sensible things here 🎉

**Comment by MathiasVP** [comment]
> If these are not causing any serious problems, I propose with just add a comment to the open issue on this that once we receive a patch we should also fix this (if that's not automatic).

I would expect this to be fixed automagically, but let's note it in the internal issue anyway 👍

**Comment by jketema** [comment]
> > If these are not causing any serious problems, I propose with just add a comment to the open issue on this that once we receive a patch we should also fix this (if that's not automatic).
> 
> I would expect this to be fixed automagically, but let's note it in the internal issue anyway 👍

Added.
